internal/controller/processor.go

```diff
@@ -172,62 +171,120 @@ func (processor *AppConfigurationProviderProcessor) ProcessKeyValueRefresh(exist
+	nextSecretReferenceRefreshReconcileTime := metav1.Time{Time: processor.CurrentTime.Add(keyVaultRefreshInterval)}
 	// When SecretSettingPopulated means ProcessFullReconciliation or ProcessKeyValueRefresh has executed, update next refresh time and return
 	if processor.RefreshOptions.SecretSettingPopulated {
-		reconcileState.NextKeyVaultReferenceRefreshReconcileTime = nextKeyVaultReferenceRefreshReconcileTime
+		reconcileState.NextSecretReferenceRefreshReconcileTime = nextSecretReferenceRefreshReconcileTime
 		return nil
 	}
 
-	resolvedSecretData, err := (*processor.Retriever).ResolveKeyVaultReferences(processor.Context, reconcileState.CachedSecretReferences, processor.ResolveSecretReference)
+	// Only resolve the secret references that not specified the secret version
+	secretReferencesToSolve := make(map[string]*loader.TargetSecretReference)
+	for secretName, reference := range reconcileState.ExistingSecretReferences {
+		for key, uriSegment := range reference.UriSegments {
+			if uriSegment.SecretVersion == "" {
+				if secretReferencesToSolve[secretName] == nil {
+					secretReferencesToSolve[secretName] = &loader.TargetSecretReference{
+						Type:        reference.Type,
+						UriSegments: make(map[string]loader.KeyVaultSecretUriSegment),
+					}
+				}
+				secretReferencesToSolve[secretName].UriSegments[key] = uriSegment
+			}
+		}
+	}
+
+	resolvedSecretData, err := (*processor.Retriever).ResolveSecretReferences(processor.Context, secretReferencesToSolve, processor.ResolveSecretReference)
 	if err != nil {
 		return err
 	}
-	maps.Copy(existingSecret.Data, resolvedSecretData)
-	processor.Settings.SecretSettings = existingSecret.Data
+
+	for secretName, targetSecret := range resolvedSecretData {
+		existingSecret, ok := existingSecrets[secretName]
+		if ok {
+			maps.Copy(existingSecret.Data, targetSecret.Data)
+		}
+
+	}
+	processor.Settings.SecretSettings = existingSecrets
 	processor.RefreshOptions.SecretSettingPopulated = true
 
 	// Update next refresh time only if settings updated successfully
-	reconcileState.NextKeyVaultReferenceRefreshReconcileTime = nextKeyVaultReferenceRefreshReconcileTime
+	reconcileState.NextSecretReferenceRefreshReconcileTime = nextSecretReferenceRefreshReconcileTime
 
 	return nil
 }
 
+func (processor *AppConfigurationProviderProcessor) shouldReconcile(
+	existingConfigMap *corev1.ConfigMap,
+	existingSecrets map[string]corev1.Secret) bool {
+
+	if processor.Provider.Generation != processor.ReconciliationState.Generation {
+		// If the provider is updated, we need to reconcile anyway
+		return true
+	}
+
+	if processor.ReconciliationState.ConfigMapResourceVersion == nil ||
+		*processor.ReconciliationState.ConfigMapResourceVersion != existingConfigMap.ResourceVersion {
+		// If the ConfigMap is removed or updated, we need to reconcile anyway
+		return true
+	}
+
+	if processor.Provider.Spec.Secret == nil {
+		return false
+	}
+
+	if len(processor.ReconciliationState.ExistingSecretReferences) == 0 ||
```

Comment:
Consider this scenario:
In beginning,  no secret-related setting being specified in the yaml. Then user updates the yaml to add Secret. In this case, `len(processor.ReconciliationState.ExistingSecretReferences) == 0` and need to reconcile

Reply:
If user updates the yaml, it can be returned early in line245. Do es this scenario happen when user updates the yaml after seeing error log? Also there's a scenario when user specified `secret` section in yaml but no key vault references in store, which may also make `len(ExistingSecretReferences)==0`

Reply:
>> Also there's a scenario when user specified secret section in yaml but no key vault references in store, which may also make len(ExistingSecretReferences)==0

No, as long as user specify the secret section, at least one secret will be generate even it's empty.

Reply:
I see, thanks

Reply:
> If user updates the yaml, it can be returned early in line245.
Right, but `len(processor.ReconciliationState.ExistingSecretReferences) == 0` is an unexpected scenario, once it happens means something was wrong(may be just because the reconciliation never succeeded)